### PR TITLE
IF: make qc_info in instant_finality_extension non-optional

### LINKED ISSUE #2128: IF: make qc_info in instant_finality_extension non-optional
From Areg in chat in reviewing https://github.com/AntelopeIO/leap/pull/2120

_I'm also wondering if we could actually get to the point where the qc_info is never optional.
For example, by leveraging the fact that we can implicitly accept a weak QC claim on the LIB._

### PR DESCRIPTION
**Pull request by greg7mdp**
Resolves #2128.

### CODE REVIEW

File: libraries/chain/block_state.cpp [hunk 1]
```diff
@@ -48,6 +50,9 @@ block_state::block_state(const block_state_legacy& bsp) {
    active_proposer_policy->active_time = bsp.timestamp();
    active_proposer_policy->proposer_schedule = bsp.active_schedule;
    header_exts = bsp.header_exts;
+
+   auto& updated_if_extension = std::get<instant_finality_extension>(header_exts.lower_bound(if_ext_id)->second);
```
Comment (heifner) [inline libraries/chain/block_state.cpp:54]: Why is this needed, why not just use:
`const auto& if_extension = std::get<instant_finality_extension>(*ext);`

Comment (greg7mdp) [reply libraries/chain/block_state.cpp:54]: This actually changes the new `header_exts` from the new `block_state` created in this constructor.
``header_exts`` is in the incoming `block_state_legacy` and I was not sure about mutating it.

Comment (greg7mdp) [reply libraries/chain/block_state.cpp:54]: I could do it otherwise. I'd have to mutate `bsp.header_exts` instead of extracting a new map in `bsp.block->extract_header_extension(if_ext_id);`.

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -3085,142 +3063,105 @@ struct controller_impl {
...
+
+      const auto& qc_ext   = std::get<quorum_certificate_extension>(block_exts.lower_bound(qc_ext_id)->second);
       const auto& qc_proof = qc_ext.qc;
 
       // Check QC information in header extension and block extension match
-      EOS_ASSERT( qc_proof.block_height == qc_claim.last_qc_block_num,
-                  block_validate_exception,
-                  "QC block number (${n1}) in block extension does not match last_qc_block_num (${n2}) in header extension. Block number: ${b}",
-                  ("n1", qc_proof.block_height)("n2", qc_claim.last_qc_block_num)("b", b->block_num()) );
+      EOS_ASSERT( qc_proof.block_num == qc_claim.last_qc_block_num,  block_validate_exception,
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/controller.cpp:3151]: I guess it is a matter of preference, I personally prefer the code to me more compact if possible, but as you originally wrote this code I'll switch back to using an extra line.

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -3085,142 +3063,105 @@ struct controller_impl {
...
+      if( !header_ext ) {
+         // If there is no header extension, ensure the block does not have a QC and also the previous
+         // block doesn't have a header extension either. Then return early.
+         // ------------------------------------------------------------------------------------------
+         EOS_ASSERT( !qc_extension_present, block_validate_exception,
+                     "Block #${b} includes a QC block extension, but doesn't have a finality header extension",
+                     ("b", block_num) );
+
+         EOS_ASSERT( !prev_header_ext, block_validate_exception,
+                     "Block #${b} doesn't have a finality header extension even though its predecessor does. Block number: ${b}",
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/controller.cpp:3088]: Good catch, fixing.

File: libraries/chain/controller.cpp [hunk 4]
```diff
@@ -3085,142 +3063,105 @@ struct controller_impl {
...
          return;
       }
 
-      const auto& if_ext = std::get<instant_finality_extension>(*header_ext);
-      if( !if_ext.qc_info ) {
-         EOS_ASSERT( block_exts.count(quorum_certificate_extension::extension_id()) == 0,
-                     block_validate_exception,
-                     "A block must have QC claim if it provides QC block extension. Block number: ${b}",
-                     ("b", b->block_num()) );
+      assert(header_ext);
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/controller.cpp:3093]: It is just a way to document (in a way that is checked at run time) that at this point in the function, we have a `header_ext`. Same as later I have `assert(header_ext && prev_header_ext);`.
Yes it is true that you can deduct it from reading the previous code, but I think this makes it nicely explicit.

File: libraries/chain/block_state.cpp (hunk 1 above)
Comment (greg7mdp) [reply libraries/chain/block_state.cpp]: It turns out that I originally added this in the wrong spot. It is now removed and we do set the `lib` in the qc_claim in the actual first transition block (see change in libraries/chain/block_header_state_legacy.cpp)

File: libraries/chain/controller.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/controller.cpp]: FWIW, I personally prefer the more compact as well.

File: .clang-format [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+AllowAllParametersOfDeclarationOnNextLine: false
+AllowShortBlocksOnASingleLine: Empty
+AllowShortFunctionsOnASingleLine: Inline
+BreakBeforeBraces: Custom
+BreakConstructorInitializers: BeforeComma
+BreakInheritanceList: BeforeComma
+ConstructorInitializerIndentWidth: 3
+ContinuationIndentWidth: 3
+MaxEmptyLinesToKeep: 3
+SortIncludes:    Never
```
Comment (heifner) [inline .clang-format:23]: At this point, I wish we had this in the repo. However, should come in a separate PR.

File: libraries/chain/controller.cpp [hunk 6]
```diff
@@ -2854,16 +2854,17 @@ struct controller_impl {
 
    static std::optional<qc_data_t> extract_qc_data(const signed_block_ptr& b) {
```
Comment (heifner) [inline libraries/chain/controller.cpp:2855]: Seems we should name this `validate_and_extract_qc_data` as it is calling the `validate_and_extract` methods of extensions.

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/controller.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/controller.cpp]: @linh2931 I'll defer to you, would you be OK with me switching back to the more compact form?

I agree that 4 lines is more legible, but basically all the `EOS_ASSERT` follow the same pattern of 4 arguments, and typically the name of the exception is not critical to understanding what the assert does, so I feel that having it more compact is beneficial, as it allows to see more of the code in one screen.

[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/controller.cpp]: Originally I crammed all into one line. As Kevin pointed out, it did not look good. I think for shorter ones, a single line is better. But for longer ones, a separate lines are better. If we do  separate lines, each argument should be on a separate line so it is easy to see.

File: .clang-format (hunk 5 above)
Comment (greg7mdp) [reply .clang-format:23]: Oh, thanks, I must have added it by mistake.

File: libraries/chain/controller.cpp (hunk 6 above)
Comment (greg7mdp) [reply libraries/chain/controller.cpp:2855]: Hum, I think it could be misleading, it does the generic extension validation, but not the actual `qc_data` validation implemented in `verify_qc_claim`.

File: libraries/chain/controller.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/controller.cpp]: I think line 1 (the condition) and line 3 (the description) are important and nice when they stand out on separate lines.

[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/controller.cpp]: > If we do separate lines, each argument should be on a separate line so it is easy to see.

I don't  agree that we sould constrain ourselves to either a single line, or as many lines as we have arguments. I feel that we should try to make the code as nice to read as possible.

[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/controller.cpp]: That's why I said shorter ones single line, longer ones separate lines. Those are all personal preferences. I look at the code in different terminals. Separate ones do look better.

**Review by linh2931 — APPROVED** [review]

**Review by heifner — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Leveraging information in previous claims, `qc_info` is always populated in the block header extension. Previously it was an optional field.
Note: end
